lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp

```diff
@@ -3064,22 +3064,41 @@ static int gdb_errno_to_system(int err) {
 
 static uint64_t ParseHostIOPacketResponse(StringExtractorGDBRemote &response,
                                           uint64_t fail_result, Status &error) {
+  // The packet is expected to have the following format:
+  // 'F<retcode>,<errno>'
+
   response.SetFilePos(0);
   if (response.GetChar() != 'F')
     return fail_result;
+
   int32_t result = response.GetS32(-2, 16);
   if (result == -2)
     return fail_result;
-  if (response.GetChar() == ',') {
-    int result_errno = gdb_errno_to_system(response.GetS32(-1, 16));
-    if (result_errno != -1)
-      error = Status(result_errno, eErrorTypePOSIX);
-    else
-      error = Status(-1, eErrorTypeGeneric);
-  } else
+
+  if (response.GetChar() != ',') {
     error.Clear();
+    return result;
+  }
+
+  // Response packet should contain a error code at the end. This code
+  // corresponds either to the gdb IOFile error code, or to the posix errno.
```

Comment:
I've taken one more look at the documentation, and there are two options that seem likely now:

1. According to the gdb rsp, error values can only be from the gdb internal list (https://sourceware.org/gdb/current/onlinedocs/gdb.html/Errno-Values.html#Errno-Values). So, in the case of ETXTBSY, the response packet value cannot contain 26 as the error code, and it must be 9999 instead, which corresponds to the unknown error. In this case, the error is most likely on the lldb server side.

2. Lldb documentation (see lldb/docs/lldb-platform-packets.txt) states that vFile:open, vFile:close, vFile:symlink, vFile:unlink, vFile:mode, and vFile:size reply packets contain errno. I think there was an attempt to extend gdb rsp so that a respond packet could contain any posix error, instead of the limited set suggested by the rsp. I think this can be possible, because as far as I know, the lldb server does not currently run on systems that are not compatible with posix.

Please let me know what you think of this matter.

Reply:
Yes it is a bit vague but going by https://lldb.llvm.org/resources/lldbgdbremote.html#vfile-open "the errno" doesn't say to me "the posix errno" necessarily. You are right that for `lldb-server` specifically, it would defacto mean the posix errno, but there is also Windows to consider.

But the code you are editing here is the client side (`lldb`). The client could be connected to a debug server running on anything.

So I don't think you can mark the unknown values `eErrorTypePOSIX` because at this point we don't know that the other end is posix. So that's why I'd go with `eErrorTypeGeneric`.

(I suspect this doesn't really matter in the end, we probably just print the error number the same way either way)

The point of your patch is to include the number instead of overwriting it with -1, and that is fine I understand why that's useful.

Reply:
One background info: `lldb/docs/lldb-platform-packets.txt` was written by me as I was reading the lldb/debugserver/lldb-server implementations of these packets, and trying to cross-reference with the official gdb remote serial protocol documentation.  Any disagreement between lldb-platform-packets.txt and the official docs should be assumed to be a mistake on my part, probably reflecting how lldb was currently handling things.  As you say, if we wanted to relay a new error type, we'd need to get gdb to add that to the "errno" table in their docs.  Otherwise the only safe error type we can use is EUNKNOWN.

Reply:
True, this would be a divergence.

I think hiding the original number is pretty annoying though, and my understanding is that's the motivation here. Even if you have access to the remote side, getting logging out of lldb-server is a pain.

We added something for adding string errors to certain packets a while ago, I think launch uses it. If I can find that, perhaps we can use it to append the original error number instead.

Reply:
https://reviews.llvm.org/D133352 / https://reviews.llvm.org/D133352

It seems to not be only for launch packets, but I was not able to confirm that.

Reply:
FWIW I think the first commit adding this support was in https://reviews.llvm.org/D34945 . lldb enables these by sending `QErrorStringInPacketSupported` to the remote stub, only after that are they sent.  debugserver currently sends error strings for qLaunchSuccess and vAttach - but I would assume that any Exx packet is allowed to include an error string if the debugger has sent `QErrorStringInPacketSupported` and it will be handled by the debugger.

Reply:
@jasonmolenda @DavidSpickett could you please confirm that my understanding is correct:
1. The error is on the lldb-server side, which have to put 9999 (EUNKNOWN) as error code instead of 26, so it should be fixed there.
2. If lldb client receives response packet with EUNKNOWN code I can try to obtain an error string from debugserver if it supports this functionality.

Reply:
> @jasonmolenda @DavidSpickett could you please confirm that my understanding is correct:
> 
> 1. The error is on the lldb-server side, which have to put 9999 (EUNKNOWN) as error code instead of 26, so it should be fixed there.

My assumption from the gdb RSP documentation you linked to earlier is that any constant not listed in that table cannot be expected to be handled correctly.  I don't know if the different errno numeric values are even specified by the POSIX doc -- they're ABI on a given platform, but I bet Darwin could use 2 for EBUSY and Linux could use 16 for EBUSY and both be compliant with the spec, so sending the errno value from the stub to lldb would not be defined.  (i'm too lazy to actually look in the POSIX spec, but all of this needs to work on non-POSIX environments on either side too, so we have to use the gdb RSP spec's documented values and only those.)

> 2. If lldb client receives response packet with EUNKNOWN code I can try to obtain an error string from debugserver if it supports this functionality.

In lldb-server if lldb has sent `QEnableErrorStrings` to enable error strings (v. `GDBRemoteCommunicationServer::Handle_QErrorStringEnable` / `GDBRemoteCommunicationServer::SendErrorResponse`), then the error code and error string (if set) will be sent to the debugger.

We should only send the errno values defined in https://sourceware.org/gdb/current/onlinedocs/gdb.html/Errno-Values.html#Errno-Values but we can send any string when lldb has enabled error strings.  The method that receives the error message may not relay it up to the user right now -- but that's a purely lldb-side issue.

Reply:
Note that the error strings, as currently used by `QEnableErrorStrings`, are only relevant for gdb protocol errors (packets of format `Exx`). The "host IO" packets discussed here use a different format for reporting errors, and we don't currently have a way to include error strings in those.

Reply:
GDB RSP supports 2 types of error response packets: `E xx` and `E.errtext` (https://sourceware.org/gdb/current/onlinedocs/gdb.html/Standard-Replies.html#Standard-Replies), so it looks like the ability to send error strings should be available for all targets by default.

Could you consider a couple of my questions about the matter, please?
1. Taking into account what I've said above, why do we need `QEnableErrorStrings`?
2. Is there any function like `SendErrorRespond`, which takes a error string and forms `E.errtext` packet? To be honest, I couldn't find something like this.

Reply:
We should have lldb send `error-message` in our QSupported list, and accept this style of error reporting from the remote stub, but that's clearly an entirely different topic than the one being discussed here.

Reply:
> Note that the error strings, as currently used by `QEnableErrorStrings`, are only relevant for gdb protocol errors (packets of format `Exx`). The "host IO" packets discussed here use a different format for reporting errors, and we don't currently have a way to include error strings in those.

Ah, excellent point, I'd forgotten that the vFile packets have a special error reporting mechanism that's entirely different. https://sourceware.org/gdb/current/onlinedocs/gdb.html/Host-I_002fO-Packets.html#Host-I_002fO-Packets

Reply:
> I think this feature was only added to the gdb remote serial protocol in May of this year?

Yeah, looks like you're right, I didn't notice this.

Thanks for your reply.